### PR TITLE
Various XRInput backend updates (+ initial Tests)

### PR DESCRIPTION
**Pull request by StayTalm**
*Previous PR Feedback
*Added initial suite of XRTests
*Exclude AraxisMerge merge files
*Added ViveLighthouse and Oculus Tracking Reference
*XR devices are now available in more XR players
*Sanitize XR feature names to be property names

### CODE REVIEW

File: Packages/com.unity.inputsystem/InputSystem/InputSystem.cs [hunk 1]
```diff
@@ -1035,10 +1035,10 @@ internal static void PerformDefaultPluginInitialization()
...
 
             #if UNITY_EDITOR || UNITY_SWITCH
-            Plugins.Switch.SwitchSupport.Initialize();
-            #endif
-
-            #if UNITY_EDITOR
+            Plugins.Switch.SwitchSupport.Initialize();
+#endif
+
+#if UNITY_EDITOR || UNITY_STANDALONE || UNITY_EDITOR || UNITY_IOS
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/InputSystem/InputSystem.cs]: Formatting

File: Packages/com.unity.inputsystem/InputSystem/Plugins/XR/GenericXRDevice.cs [hunk 2]
```diff
@@ -89,6 +88,14 @@ public void SetIntensity(float intensity)
             m_Rumble.intensity = intensity;
         }
 
+        public bool IsHapticsPaused
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/XR/GenericXRDevice.cs]: Lowercase

File: Packages/com.unity.inputsystem/InputSystem/Plugins/XR/Vive.cs [hunk 3]
```diff
@@ -89,4 +89,23 @@ protected override void FinishSetup(InputDeviceBuilder builder)
             deviceAngularVelocity = builder.GetControl<Vector3Control>("deviceAngularVelocity");
         }
     }
+
+    [InputControlLayout]
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/InputSystem/Plugins/XR/Vive.cs:93]: Technically you don't even need these. When registering a type as a layout, it takes the type whether the attribute is there or not. The attribute just allows controlling some settings on the layout.

Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/InputSystem/Plugins/XR/Vive.cs:93]: I don't mind them being there, though. Kinda makes it a little more explicit.

File: Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs [hunk 4]
```diff
@@ -1,29 +1,648 @@
...
+                        featureType = FeatureType.Binary
+                    }
+                }
+            }.ToJson()
+        }.ToJson());
+
+        InputSystem.Update();
+
+        var generatedLayout = InputSystem.TryLoadLayout("XRInput::XRManufacturer::XRGenericDevice");
+        Assert.That(generatedLayout, Is.Not.EqualTo(null));
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs]: Alternatively there's `Is.Not.Null`.

File: Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs [hunk 5]
```diff
@@ -1,29 +1,648 @@
...
+                inputFeatures = new List<XRFeatureDescriptor>()
+                {
+                    new XRFeatureDescriptor()
+                    {
+                        name = "SimpleFeature",
+                        featureType = FeatureType.Binary
+                    }
+                }
+            }.ToJson()
+        }.ToJson());
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs]: Might be worth building something equivalent to `HID.HIDDeviceDescriptorBuilder` for XR.

File: Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs [hunk 6]
```diff
@@ -1,29 +1,648 @@
...
+
+    [Test]
+    [Category("Devices")]
+    public void Devices_HardwareTrackerCreatesDefaultInputDevice()
+    {
+        testRuntime.ReportNewInputDevice(
+            new InputDeviceDescription
+        {
+            interfaceName = XRUtilities.kXRInterface,
+            product = "HardwareTracker",
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs]: What's a "HardwareTracker"?

File: Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs [hunk 7]
```diff
@@ -1,29 +1,648 @@
...
+        });
+        InputSystem.Update();
+
+        Assert.That((device["Button"] as ButtonControl).isPressed, Is.EqualTo(false));
+        Assert.That(device["DiscreteState"].ReadValueAsObject(), Is.EqualTo(0));
+        Assert.That(device["Axis"].ReadValueAsObject(), Is.EqualTo(0f).Within(0.0001f));
+        Assert.That(device["Vector2"].ReadValueAsObject(), Is.EqualTo(Vector2.zero));
+        Assert.That(device["Vector3"].ReadValueAsObject(), Is.EqualTo(Vector3.zero));
+        Assert.That(device["Rotation"].ReadValueAsObject(), Is.EqualTo(Quaternion.identity));
+        Assert.That(device["Custom"], Is.EqualTo(null));
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs]: `Is.Null`

File: Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs [hunk 8]
```diff
@@ -1,29 +1,648 @@
...
+        Assert.That((device["Last"] as ButtonControl).isPressed, Is.EqualTo(false));
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs]: `Is.False`

File: Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs [hunk 9]
```diff
@@ -1,29 +1,648 @@
...
+            discreteState = 17,
+            axis = 1.24f,
+            axis2D = new Vector2(0.1f, 0.2f),
+            axis3D = new Vector3(0.3f, 0.4f, 0.5f),
+            rotation = new Quaternion(0.6f, 0.7f, 0.8f, 0.9f),
+            lastElement = byte.MaxValue,
+        });
+        InputSystem.Update();
+
+        Assert.That((device["Button"] as ButtonControl).isPressed, Is.EqualTo(true));
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs]: `Is.True`

File: Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs [hunk 10]
```diff
@@ -1,29 +1,648 @@
...
+        var generatedLayout = InputSystem.TryLoadLayout("XRInput::XRManufacturer::XRDevice1");
+        Assert.That(generatedLayout, Is.EqualTo(null));
+
+        generatedLayout = InputSystem.TryLoadLayout("XRInput::XRManufacturer::XRDevice2");
+        Assert.That(generatedLayout, Is.EqualTo(null));
+    }
+
+    [Test]
+    [Category("Layouts")]
+    public void Layouts_WhenMatchingAKnownDeviceTheGeneratedLayoutInheritsProperly()
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs]: Avoid "proper" and "correct" in test names and try to be specific instead. Also, this test seems to single out one case with no guarantee that the other cases work the same. Recommend converting this test into one that takes the product+manufacturer combination as well as the base layout name as a parameter and then uses `[TestCase(...)]`.

File: Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs [hunk 11]
```diff
@@ -1,29 +1,648 @@
...
+            {
+                deviceRole = DeviceRole.LeftHanded,
+                inputFeatures = new List<XRFeatureDescriptor>()
+                {
+                    new XRFeatureDescriptor()
+                    {
+                        name = "SimpleFeature",
+                        featureType = FeatureType.Binary
+                    }
+                }
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs]: This feature doesn't seem relevant to the test. Same for some other tests here.

File: Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs (hunk 10 above)
Comment (Rene-Damm) [reply Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs]: Also recommend breaking up test names when there's a condition+outcome kinda setup. I.e. Devices_WhenThisIsTheCase_ThenThisIsTheResult. Just helps navigating test names a bit.

File: Packages/com.unity.inputsystem/Tests/InputSystem/UnitTests.meta [hunk 12]
```diff
@@ -0,0 +1,8 @@
+fileFormatVersion: 2
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/Tests/InputSystem/UnitTests.meta]: Please remove this file. The UnitTests folder has been removed.

**Review by Rene-Damm — CHANGES_REQUESTED** [review]

File: Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs (hunk 6 above)
Comment (StayTalm) [reply Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs]: A hardware tracker is the hockey puck device we have that is attached to objects to render in game.  https://www.vive.com/us/vive-tracker/

File: Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs (hunk 11 above)
Comment (StayTalm) [reply Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs]: I need a device state of at least 1 element, since it needs to have a size > 0 in the device state.
Although, that reminds me... I should have a test and feature that excludes devices of 0 features.

File: Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs (hunk 5 above)
Comment (StayTalm) [reply Packages/com.unity.inputsystem/Tests/InputSystem/Plugins/XRTests.cs]: Playing around with the XRDeviceDescriptorBuilder, and I'm finding it's about the same amount of lines to build, but now needs some whole other class.  I'm thinking of holding off on that one for now, since testing is the only time I'd really create a XRDeviceDescriptor in C#.

**Comment by StayTalm** [comment]
@Rene-Damm  Updated tests and took feedback into account.  Hoping to leave XML Docs on this round, and tackle them on the next pass.

**Review by Rene-Damm — APPROVED** [review]
